<div class="post-text" itemprop="text">
<p>I am currently writing a program in Python 3.3.0 which prints out the n first square numbers and finally prints out their sum. The condition is that the user can only compute an integral number of terms greater than zero. Here is the code:</p>
<pre><code>print("WELCOME!")

n = input("How many numbers to sum up?: ")

while n &lt;= 0:
    print("You have to write a positive integer; try again!")
    n = input("How many numbers to sum up?: ")

i = 1
sum = 0
while 0 &lt; i &lt;= n:
    print(i,"*",i,"=", i**2)
    sum += i**2
    i += 1
print("Sum:", sum)
</code></pre>
<p>Now, I managed to code the error message for when the user types in a negative number. However, I am having difficulties coding an error message for when the user types in a string, such as "fifteen". </p>
<p>I want the program to allow the user to try assigning 'n' a value once again and again, just as if they had typed a negative number; basically I want to code a loop asking for the same thing over and over again if the user inputs a string.</p>
<p>The problem is that input() always assigns a string to an arbitrary variable, so I tried writing code for converting the string to an integer. It works fine when the user inputs an integral value, however it cannot define int(n) if 'n' is not an integer.</p>
<p>I have googled a lot regarding this issue and I found some examples using try and except ValueError but none of them seem to be able to create a loop out of it.</p>
<p>Anybody got an idea?</p>
</div>
<div class="post-text" itemprop="text">
<p>Put the <code>try/except</code> inside the loop.</p>
<pre><code>while True:
    try:
        n = int(input("How many numbers to sum up?: "))
        if n &lt;= 0:
            print("You have to write a positive integer; try again!")
            continue
        break
    except ValueError:
        print("You have to write a positive integer; try again!")
</code></pre>
<p>You could also replace the last part of your code (the part after the first loop) with this:</p>
<pre><code>squares = lambda x: [print('{0} * {0} = {1}'.format(x, x**2)), x**2][1]

print("Sum:", sum(squares(i) for i in range(1, n+1))
</code></pre>
<p>The lambda isn't recommended though <code>;)</code>. You could of course use a normal function instead:</p>
<pre><code>def squares(x):
    print('{0} * {0} = {1}'.format(x, x**2))
    return x**2
</code></pre>
</div>
<div class="post-text" itemprop="text">
<p>You can use <a href="http://docs.python.org/3/library/stdtypes.html#str.isdigit" rel="nofollow">str.isdigit()</a> to accomplish this.  According to the docs, isdigit() accepts compatibility superscript digits such as \u0660123.  But python 3's input() escapes the backslash for this type of input so these strings shouldn't cause any issues.</p>
<pre><code>while True:
  n = input("How many numbers to sum up?: ")
  if n.isdigit() and int(n) &gt; 0:
    break
  else:
    print("You have to write a positive integer; try again!")

squares = (i**2 for i in range(1,int(n)+1))
s = 0
for integer in range(1,int(n)+1):
  nextsquare = next(squares)
  s += nextsquare
  print(integer, "*", integer, '=', nextsquare, sep='')

print('Sum', s)
</code></pre>
</div>
<div class="post-text" itemprop="text">
<p>I guess something like this? pseudo-python:</p>
<pre><code>n = None
while n is None:
    try:
        n = int(input("How many numbers to sum up?: "))
    except ValueError:
        pass
</code></pre>
</div>
<span class="comment-copy">You may also have a look at <a href="http://docs.python.org/2/library/ast.html#ast.literal_eval" rel="nofollow noreferrer"><code>ast.literal_eval</code></a></span>
<span class="comment-copy">Thank you, that did the trick.</span>
